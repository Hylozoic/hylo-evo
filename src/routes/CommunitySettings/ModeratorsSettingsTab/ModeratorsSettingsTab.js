import PropTypes from 'prop-types'
import React, { Component } from 'react'
import styles from './ModeratorsSettingsTab.scss'
import Loading from 'components/Loading'
import { KeyControlledItemList } from 'components/KeyControlledList'
import RemovableListItem from 'components/RemovableListItem'
import { isEmpty, get } from 'lodash/fp'
import { getKeyCode, keyMap } from 'util/textInput'

const { array, func, string } = PropTypes

export default class ModeratorsSettingsTab extends Component {
  static propTypes = {
    moderators: array,
    removeModerator: func,
    addModerator: func,
    slug: string
  }

  componentWillUnmount () {
    this.props.clearModeratorSuggestions()
  }

  render () {
    const {
      moderators,
      removeModerator,
      fetchModeratorSuggestions,
      addModerator,
      moderatorSuggestions,
      clearModeratorSuggestions
    } = this.props

    if (!moderators) return <Loading />

    return <div>
      <div>
        {moderators.map(m =>
          <RemovableListItem
            item={m}
            removeModerator={removeModerator}
            confirmMessage={`Are you sure you want to remove ${m.name}'s moderator powers?`}
            key={m.id} />)}
      </div>
      <AddModerator
        fetchModeratorSuggestions={fetchModeratorSuggestions}
        addModerator={addModerator}
        moderatorSuggestions={moderatorSuggestions}
        clearModeratorSuggestions={clearModeratorSuggestions} />
    </div>
  }
}

export class AddModerator extends Component {
  static propTypes = {
    addModerator: func,
    fetchModeratorSuggestions: func
  }

  constructor (props) {
    super(props)
    this.state = {
      adding: false
    }
  }

  render () {
    const { fetchModeratorSuggestions, addModerator, moderatorSuggestions, clearModeratorSuggestions } = this.props

    const { adding } = this.state

    const toggle = () => {
      this.setState({adding: !adding})
    }

    const onInputChange = e => {
      if (e.target.value.length === 0) return clearModeratorSuggestions()
      return fetchModeratorSuggestions(e.target.value)
    }

    const onChoose = choice => {
      addModerator(choice.id)
      clearModeratorSuggestions()
      toggle()
    }

    const chooseCurrentItem = () => {
      if (!this.refs.list) return
      return this.refs.list.handleKeys({
        keyCode: keyMap.ENTER,
        preventDefault: () => {}
      })
    }

    const handleKeys = e => {
      if (getKeyCode(e) === keyMap.ESC) {
        toggle()
        return clearModeratorSuggestions()
      }
      if (!this.refs.list) return
      return this.refs.list.handleKeys(e)
    }

    const listWidth = {width: get('refs.input.clientWidth', this, 0) + 4}

    if (adding) {
      return <div styleName='add-moderator adding'>
        <div styleName='help-text'>Search here for members to grant moderator powers</div>
        <div styleName='input-row'>
          <input styleName='input'
            placeholder='Type...'
            type='text'
            onChange={onInputChange}
            onKeyDown={handleKeys}
            ref='input' />
          <span styleName='cancel-button' onClick={toggle}>Cancel</span>
          <span styleName='add-button' onClick={chooseCurrentItem}>Add</span>
        </div>
        {!isEmpty(moderatorSuggestions) && <div style={listWidth}>
          <KeyControlledItemList
            ref='list'
            items={moderatorSuggestions}
            onChange={onChoose}
            theme={styles} />
        </div>}
      </div>
    } else {
      return <div styleName='add-moderator add-new' onClick={toggle}>
        + Add New
      </div>
    }
  }
}
