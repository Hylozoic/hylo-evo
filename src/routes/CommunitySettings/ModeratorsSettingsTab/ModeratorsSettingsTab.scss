// AddModerator
.add-moderator {
  margin-top: 15px;
}

.adding {
  width: 100%;
}

.help-text {
  composes: caption-lt-lg from 'css/typography.scss';
  margin-bottom: 10px;
}

.input-row {
  display: flex;
  align-items: center;
}

.input {
  flex: 1;
  margin-right: 10px;
  padding: 5px 10px;
  color: $color-rhino;
  &::placeholder {
    color: $color-rhino-60;
  }
}

.cancel-button {
  composes: text-button from 'css/typography.scss';
  color: $color-rhino-60;
  &:hover {
    color: $color-rhino-80;
  }
  margin-right: 10px;
}

.add-button {
  composes: text-button from 'css/typography.scss';
}

.add-new {
  height: 30px;
  line-height: 30px;
  text-align: center;
  border: 1px dashed $color-ghost;
  composes: text-button from 'css/typography.scss'
}

// KeyControlledListTheme

 .items {
   list-style: none;
   border: 1px solid $color-ghost;
   padding: 0;
   li {
     border-top: 1px solid $color-ghost;
     &:first-child {
       border: none;
     }
   }
 }

.item {
  padding: 5px 5px 5px 11px;
  color: $color-caribbean-green;
}

.item-active {
  background-color: $color-athens-gray-dark;
}
