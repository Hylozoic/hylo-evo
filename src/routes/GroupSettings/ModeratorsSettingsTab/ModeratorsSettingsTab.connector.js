import { connect } from 'react-redux'
import {
  fetchModeratorSuggestions,
  clearModeratorSuggestions,
  getModerators,
  addModerator,
  removeModerator
} from './ModeratorsSettingsTab.store'
import {
  addGroupRole,
  addRoleToMember,
  fetchMembersForGroupRole,
  removeRoleFromMember,
  updateGroupRole
} from '../../../store/actions/roles'
import getPerson from 'store/selectors/getPerson'
import { includes } from 'lodash/fp'

export function mapStateToProps (state, props) {
  const moderators = getModerators(state, props)
  const moderatorIds = moderators.map(m => m.id)
  const moderatorSuggestions = state.ModeratorsSettings
    .filter(personId => !includes(personId, moderatorIds))
    .map(personId => getPerson(state, { personId }))

  const rawSuggestions = state.ModeratorsSettings
    .map(personId => getPerson(state, { personId }))

  return {
    moderators,
    moderatorSuggestions,
    rawSuggestions
  }
}

export function mapDispatchToProps (dispatch, props) {
  const { groupId } = props

  return {
    addModerator: id => dispatch(addModerator(id, groupId)),
    addGroupRole: params => dispatch(addGroupRole(params)),
    addRoleToMember: params => dispatch(addRoleToMember({ ...params, groupId })),
    clearModeratorSuggestions: () => dispatch(clearModeratorSuggestions()),
    fetchModeratorSuggestions: autocomplete => dispatch(fetchModeratorSuggestions(groupId, autocomplete)),
    fetchMembersForGroupRole: params => dispatch(fetchMembersForGroupRole({ ...params, id: groupId })),
    removeModerator: (id, isRemoveFromGroup) => dispatch(removeModerator(id, groupId, isRemoveFromGroup)),
    removeRoleFromMember: params => dispatch(removeRoleFromMember({ ...params, groupId })),
    updateGroupRole: params => dispatch(updateGroupRole(params))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)
