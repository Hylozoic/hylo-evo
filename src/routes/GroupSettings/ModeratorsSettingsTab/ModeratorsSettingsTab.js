import PropTypes from 'prop-types'
import React, { Component, useState, useEffect } from 'react'
import Loading from 'components/Loading'
import Icon from 'components/Icon'
import { KeyControlledItemList } from 'components/KeyControlledList'
import RemovableListItem from 'components/RemovableListItem'
import { isEmpty, get, includes } from 'lodash/fp'
import { getKeyCode, keyMap } from 'util/textInput'
import { personUrl } from 'util/navigation'
import SettingsControl from 'components/SettingsControl'
import SettingsSection from '../SettingsSection'
import ModalDialog from 'components/ModalDialog'
import CheckBox from 'components/CheckBox'
import EmojiPicker from 'components/EmojiPicker'
import general from '../GroupSettings.scss' // eslint-disable-line no-unused-vars
import styles from './ModeratorsSettingsTab.scss' // eslint-disable-line no-unused-vars

const { array, func, string, object } = PropTypes

const emptyRole = {
  color: '',
  description: '',
  emoji: '',
  name: '',
  active: ''
}

const validateRole = ({ name, emoji }) => {
  if (name.length < 3) return false
  if (emoji === '') return false
  return true
}

export default class ModeratorsSettingsTab extends Component {
  static propTypes = {
    addModerator: func,
    addGroupRole: func,
    addRoleToMember: func,
    group: object,
    moderators: array,
    removeModerator: func,
    removeRoleFromMember: func,
    roles: array,
    slug: string,
    updateGroupRole: func
  }

  constructor (props) {
    super(props)
    this.state = this.defaultEditState()
  }

  state = {
    modalVisible: false
  }

  componentDidUpdate (prevProps, prevState) {
    if (!prevProps.group?.groupRoles && this.props.group?.groupRoles) {
      this.setState(this.defaultEditState())
    }
  }

  componentWillUnmount () {
    this.props.clearModeratorSuggestions()
  }

  defaultEditState () {
    const { group } = this.props

    if (!group) return { roles: [], valid: false }

    const roles = group?.groupRoles?.items

    return {
      roles: roles || [],
      error: null,
      modalVisible: false
    }
  }

  submitRemoveModerator = () => {
    this.props.removeModerator(this.state.moderatorToRemove, this.state.isRemoveFromGroup)
  }

  handleAddRole = () => {
    this.setState({
      roles: [...this.state.roles].concat({ ...emptyRole })
    })
  }

  deleteUnsavedRole = (i) => () => {
    if (window.confirm('Are you sure you want to delete this unsaved role/badge?')) { // TODO: i18n
      const newRoles = [...this.state.roles]
      newRoles.splice(i, 1)
      this.setState({
        roles: newRoles
      })
    }
  }

  toggleRoleActivation = (i) => () => {
    const roles = [...this.state.roles]
    const role = roles[i]
    if (window.confirm(`Are you sure you want to ${role.active ? 'deactivate' : 'reactivate'} this role/badge?`)) { // TODO: i18n
      this.props.updateGroupRole({ active: !role.active, groupId: this.props?.group?.id, groupRoleId: role.id }).then((response) => {
        roles[i] = { ...response.payload.data.updateGroupRole }
        this.setState({ roles })
      })
    }
  }

  updateLocalRole = (i) => (key) => (v) => {
    const value = typeof (v.target) !== 'undefined' ? v.target.value : v
    const role = { ...this.state.roles[i] }
    if (role.changed !== true) role.originalState = { ...role }
    role[key] = value
    role.changed = true
    const roles = [...this.state.roles]
    roles[i] = role
    this.setState({ roles })
  }

  saveRole = (i) => () => {
    const role = { ...this.state.roles[i] }
    if (validateRole(role)) {
      this.props.addGroupRole({ ...role, groupId: this.props?.group?.id }).then((response) => {
        const roles = this.state.roles
        roles[i] = { ...response.payload.data.addGroupRole }
        this.setState({ roles })
      })
    } else {
      window.alert('A role must have a valid emoji and name to be saved') // TODO: i18n
    }
  }

  resetRole = (i) => () => {
    const role = { ...this.state.roles[i] }
    const roles = [...this.state.roles]
    roles[i] = { ...role.originalState }
    this.setState({ roles })
  }

  updateRole = (i) => () => {
    const role = { ...this.state.roles[i] }
    if (validateRole(role)) {
      this.props.updateGroupRole({ ...role, groupId: this.props?.group?.id, groupRoleId: role.id }).then((response) => {
        const roles = this.state.roles
        roles[i] = { ...response.payload.data.updateGroupRole }
        this.setState({ roles })
      })
    } else {
      window.alert('A role must have a valid emoji and name to be updated') // TODO: i18n
    }
  }

  render () {
    const {
      moderators,
      group = {}
    } = this.props

    const {
      modalVisible,
      isRemoveFromGroup,
      roles
    } = this.state

    const unsavedRolePresent = roles.length > 0 ? roles[roles.length - 1]?.active === '' : false
    if (!moderators) return <Loading />
    return (
      <>
        <SettingsSection>
          <h3>
            {group?.moderatorDescriptorPlural || 'Moderators'}
            <div styleName='styles.help-text'>Who has access to group settings and moderation powers</div>
            {/* TODO i18n */}
          </h3>
          <ModeratorsList key='mList' {...this.props} removeItem={(id) => this.setState({ modalVisible: true, moderatorToRemove: id })} />
          {modalVisible && <ModalDialog
            key='remove-moderator-dialog'
            closeModal={() => this.setState({ modalVisible: false })}
            showModalTitle={false}
            submitButtonAction={this.submitRemoveModerator}
            submitButtonText='Remove'>
            <div styleName='styles.content'>
              <div styleName='styles.modal-text'>Are you sure you wish to remove this moderator?</div>
              <CheckBox checked={isRemoveFromGroup} label='Remove from group as well' onChange={value => this.setState({ isRemoveFromGroup: value })} />
            </div>
          </ModalDialog>}
        </SettingsSection>
        <SettingsSection>
          <h3>Other Roles &amp; Badges</h3>
          <div styleName='styles.help-text'>Create additional roles or badges for the group</div>
          {roles.map((role, i) => (
            <RoleRow
              {...this.props}
              group={group}
              key={i}
              index={i}
              {...role}
              onChange={this.updateLocalRole(i)}
              onSave={this.saveRole(i)}
              onUpdate={this.updateRole(i)}
              onToggleActivation={this.toggleRoleActivation(i)}
              onDelete={this.deleteUnsavedRole(i)}
              onReset={this.resetRole(i)}
            />
          ))}
          {!unsavedRolePresent && (
            <div styleName='styles.add-role' onClick={this.handleAddRole}>
              <h4>Create new role/badge</h4>
              <Icon name='Circle-Plus' styleName='styles.new-role' />
            </div>
          )}
        </SettingsSection>
        <br />
      </>
    )
  }
}

export function ModeratorsList ({ moderators, slug, removeItem, fetchModeratorSuggestions, addModerator, moderatorSuggestions, clearModeratorSuggestions }) {
  return (
    <div>
      <div>
        {moderators.map(m =>
          <RemovableListItem
            item={m}
            url={personUrl(m.id, slug)}
            skipConfirm
            removeItem={removeItem}
            key={m.id}
          />)}
      </div>
      <AddModerator
        fetchModeratorSuggestions={fetchModeratorSuggestions}
        addModerator={addModerator}
        moderatorSuggestions={moderatorSuggestions}
        clearModeratorSuggestions={clearModeratorSuggestions}
      />
    </div>
  )
}

export class AddModerator extends Component {
  static propTypes = {
    addModerator: func,
    fetchModeratorSuggestions: func
  }

  constructor (props) {
    super(props)
    this.state = {
      adding: false
    }
  }

  render () {
    const { fetchModeratorSuggestions, addModerator, moderatorSuggestions, clearModeratorSuggestions } = this.props

    const { adding } = this.state

    const toggle = () => {
      this.setState({ adding: !adding })
    }

    const onInputChange = e => {
      if (e.target.value.length === 0) return clearModeratorSuggestions()
      return fetchModeratorSuggestions(e.target.value)
    }

    const onChoose = choice => {
      addModerator(choice.id)
      clearModeratorSuggestions()
      toggle()
    }

    const chooseCurrentItem = () => {
      if (!this.refs.list) return
      return this.refs.list.handleKeys({
        keyCode: keyMap.ENTER,
        preventDefault: () => {}
      })
    }

    const handleKeys = e => {
      if (getKeyCode(e) === keyMap.ESC) {
        toggle()
        return clearModeratorSuggestions()
      }
      if (!this.refs.list) return
      return this.refs.list.handleKeys(e)
    }

    const listWidth = { width: get('refs.input.clientWidth', this, 0) + 4 }

    if (adding) {
      return (
        <div styleName='styles.adding'>
          <div styleName='styles.help-text'>Search here for members to grant moderator powers</div>
          <div styleName='styles.input-row'>
            <input
              styleName='styles.input'
              placeholder='Type...'
              type='text'
              onChange={onInputChange}
              onKeyDown={handleKeys}
              ref='input'
            />
            <span styleName='styles.cancel-button' onClick={toggle}>Cancel</span>
            <span styleName='styles.add-button' onClick={chooseCurrentItem}>Add</span>
          </div>
          {!isEmpty(moderatorSuggestions) && <div style={listWidth}>
            <KeyControlledItemList
              ref='list'
              items={moderatorSuggestions}
              onChange={onChoose}
              theme={styles}
            />
          </div>}
        </div>
      )
    } else {
      return (
        <div styleName='styles.add-new' onClick={toggle}>
          + Add New
        </div>
      )
    }
  }
}

function RoleRow ({
  active,
  addRoleToMember,
  changed,
  clearModeratorSuggestions,
  description,
  emoji,
  fetchModeratorSuggestions,
  fetchMembersForGroupRole,
  group,
  id,
  name,
  onChange,
  onDelete,
  onToggleActivation,
  onReset,
  onSave,
  onUpdate,
  index,
  rawSuggestions = [],
  removeRoleFromMember
}) {
  const isDraftRole = active === ''
  const inactiveStyle = (!active && !isDraftRole) ? 'styles.inactive' : ''
  return (
    <div styleName={`styles.role-container ${inactiveStyle}`}>
      <div styleName='styles.action-container'>
        {isDraftRole && (<span onClick={onDelete} styleName='styles.action'><Icon name='Trash' /> Delete</span>)}
        {!isDraftRole && changed && (<span styleName='styles.action' onClick={onUpdate}><Icon name='Unlock' /> Save</span>)}
        {!isDraftRole && changed && (<span styleName='styles.action' onClick={onReset}><Icon name='Back' /> Revert</span>)}
        {!isDraftRole && !changed && (<span styleName='styles.action' onClick={onToggleActivation}><Icon name={active ? 'CircleEx' : 'CircleArrow'} /> {active ? 'Deactivate' : 'Reactivate'}</span>)}
        {/* TODO: i18n */}
      </div>
      <div styleName='styles.role-row'>
        <EmojiPicker forReactions={false} emoji={emoji} handleReaction={onChange('emoji')} className={styles['emoji-picker']} />
        <div styleName='styles.role-stack'>
          <SettingsControl label='Name' controlClass={styles['settings-control']} onChange={onChange('name')} value={name} />
          <SettingsControl label='Description' controlClass={styles['settings-control']} onChange={onChange('description')} value={description} />
        </div>
      </div>
      {
        isDraftRole
          ? (
            <div styleName='styles.role-row styles.reverse-flex'>
              <div styleName='styles.create-button' onClick={onSave}>Create Role</div>
            </div>
          )
          : active && (
            <SettingsSection>
              <GroupRoleList
                {...{ addRoleToMember, rawSuggestions, clearModeratorSuggestions, fetchMembersForGroupRole, fetchModeratorSuggestions, removeRoleFromMember }}
                key='grList'
                groupRoleId={id}
                slug={group.slug}
              />
            </SettingsSection>
          )
      }
    </div>
  )
}

export class AddMemberToRole extends Component {
  static propTypes = {
    addRoleToMember: func,
    clearSuggestions: func,
    fetchSuggestions: func,
    groupRoleId: string,
    memberSuggestions: array,
    updateLocalMembersForRole: func
  }

  constructor (props) {
    super(props)
    this.state = {
      adding: false
    }
  }

  render () {
    const { fetchSuggestions, addRoleToMember, memberSuggestions, clearSuggestions, groupRoleId, updateLocalMembersForRole } = this.props

    const { adding } = this.state

    const toggle = () => {
      clearSuggestions()
      this.setState({ adding: !adding })
    }

    const onInputChange = e => {
      if (e.target.value.length === 0) return clearSuggestions()
      return fetchSuggestions(e.target.value)
    }

    const onChoose = choice => {
      addRoleToMember({ personId: choice.id, groupRoleId }).then(() => {
        updateLocalMembersForRole(choice)
      })
      toggle()
    }

    const chooseCurrentItem = () => {
      if (!this.refs.list) return
      return this.refs.list.handleKeys({
        keyCode: keyMap.ENTER,
        preventDefault: () => {}
      })
    }

    const handleKeys = e => {
      if (getKeyCode(e) === keyMap.ESC) {
        toggle()
        return
      }
      if (!this.refs.list) return
      return this.refs.list.handleKeys(e)
    }

    const listWidth = { width: get('refs.input.clientWidth', this, 0) + 4 }

    if (adding) {
      return (
        <div styleName='styles.adding'>
          <div styleName='styles.help-text'>Search here for members to grant this role too</div>
          {/* TODO: i18n */}
          <div styleName='styles.input-row'>
            <input
              styleName='styles.input'
              placeholder='Type...'
              type='text'
              onChange={onInputChange}
              onKeyDown={handleKeys}
              ref='input'
            />
            <span styleName='styles.cancel-button' onClick={toggle}>Cancel</span>
            <span styleName='styles.add-button' onClick={chooseCurrentItem}>Add</span>
          </div>
          {!isEmpty(memberSuggestions) && <div style={listWidth}>
            <KeyControlledItemList
              ref='list'
              items={memberSuggestions}
              onChange={onChoose}
              theme={styles}
            />
          </div>}
        </div>
      )
    } else {
      // TODO: i18n
      return (
        <div styleName='styles.add-new' onClick={toggle}>
          + Add Member to Role
        </div>
      )
    }
  }
}

export function GroupRoleList ({ slug, removeItem, fetchModeratorSuggestions, addRoleToMember, rawSuggestions, clearModeratorSuggestions, groupRoleId, fetchMembersForGroupRole, removeRoleFromMember }) {
  const [membersForRole, setMembersForRole] = useState([])

  useEffect(() => {
    fetchMembersForGroupRole({ groupRoleId })
      .then((response) => {
        setMembersForRole(response.payload.data.group.members.items)
      })
  }, [])

  const memberRoleIds = membersForRole.map(mr => mr.id)

  const memberSuggestions = rawSuggestions.filter(person => !includes(person.id, memberRoleIds))

  const updateLocalMembersForRole = (choice) => {
    const updatedMembers = [...membersForRole, choice]
    setMembersForRole(updatedMembers)
  }

  const handleRemoveRoleFromMember = (id) => {
    removeRoleFromMember({ personId: id, groupRoleId }).then(() => {
      const updatedMembers = membersForRole.filter(member => member.id !== id)
      setMembersForRole(updatedMembers)
    })
  }

  return (
    <div>
      <div>
        {membersForRole.map(m =>
          <RemovableListItem
            item={m}
            url={personUrl(m.id, slug)}
            removeItem={handleRemoveRoleFromMember}
            key={m.id}
          />)}
      </div>
      <AddMemberToRole
        fetchSuggestions={fetchModeratorSuggestions}
        addRoleToMember={addRoleToMember}
        memberSuggestions={memberSuggestions}
        clearSuggestions={clearModeratorSuggestions}
        updateLocalMembersForRole={updateLocalMembersForRole}
        groupRoleId={groupRoleId}
      />
    </div>
  )
}
