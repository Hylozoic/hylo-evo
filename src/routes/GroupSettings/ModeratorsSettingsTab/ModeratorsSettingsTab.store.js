import { createSelector as ormCreateSelector } from 'redux-orm'
import orm from 'store/models'
import gql from 'graphql-tag'
import {
  CLEAR_MODERATOR_SUGGESTIONS,
  FETCH_MODERATOR_SUGGESTIONS,
  ADD_MODERATOR,
  REMOVE_MODERATOR
} from 'store/constants'

export const MODULE_NAME = 'ModeratorsSettingsTab'

const defaultState = []

export default function reducer (state = defaultState, action) {
  const { error, type, payload } = action
  if (error) return state

  switch (type) {
    case FETCH_MODERATOR_SUGGESTIONS:
      return payload.data.group.members.items.map(m => m.id)
    case CLEAR_MODERATOR_SUGGESTIONS:
      return []
    default:
      return state
  }
}

export function fetchModeratorSuggestions (id, autocomplete) {
  return {
    type: FETCH_MODERATOR_SUGGESTIONS,
    graphql: {
      query: gql`
        query ($id: ID, $autocomplete: String) {
          group (id: $id) {
            id
            members (first: 10, autocomplete: $autocomplete) {
              hasMore
              items {
                id
                name
                avatarUrl
              }
            }
          }
        }
      `,
      variables: {
        id, autocomplete
      }
    },
    meta: {
      extractModel: 'Group'
    }
  }
}

export function clearModeratorSuggestions () {
  return {
    type: CLEAR_MODERATOR_SUGGESTIONS
  }
}

export function addModerator (personId, groupId) {
  return {
    type: ADD_MODERATOR,
    graphql: {
      query: gql`
        mutation ($personId: ID, $groupId: ID) {
          addModerator(personId: $personId, groupId: $groupId) {
            id
            moderators (first: 100) {
              items {
                id
                name
                avatarUrl
              }
            }
          }
        }
      `,
      variables: { personId, groupId }
    },
    meta: {
      personId,
      groupId,
      optimistic: true
    }
  }
}

export function removeModerator (personId, groupId, isRemoveFromGroup) {
  return {
    type: REMOVE_MODERATOR,
    graphql: {
      query: gql`
        mutation ($personId: ID, $groupId: ID, $isRemoveFromGroup: Boolean) {
          removeModerator(personId: $personId, groupId: $groupId, isRemoveFromGroup: $isRemoveFromGroup) {
            id
            moderators (first: 100) {
              items {
                id
                name
                avatarUrl
              }
            }
          }
        }
      `,
      variables: { personId, groupId, isRemoveFromGroup }
    },
    meta: {
      personId,
      groupId,
      isRemoveFromGroup,
      optimistic: true
    }
  }
}

// expects props to be of the form {groupId}
export const getModerators = ormCreateSelector(
  orm,
  (state, props) => props.groupId,
  ({ Group }, id) => {
    const group = Group.safeGet({ id })
    if (!group) return []
    return group.moderators.toModelArray()
  }
)
