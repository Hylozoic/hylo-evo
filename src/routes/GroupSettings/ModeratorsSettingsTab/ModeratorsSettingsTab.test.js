import React from 'react'
import { shallow } from 'enzyme'
import ModeratorsSettingsTab, { AddModerator, ModeratorsList } from './ModeratorsSettingsTab'
import { keyMap } from 'util/textInput'

jest.mock('react-i18next', () => ({
  ...jest.requireActual('react-i18next'),
  useTranslation: (domain) => {
    return {
      t: (str) => str,
      i18n: {
        changeLanguage: () => new Promise(() => {})
      }
    }
  },
  withTranslation: () => Component => {
    Component.defaultProps = { ...Component.defaultProps, t: () => '' }
    return Component
  }
}))

describe('ModeratorsSettingsTab', () => {
  it('renders loading if no moderators', () => {
    const wrapper = shallow(<ModeratorsSettingsTab />)
    expect(wrapper).toMatchSnapshot()
  })

  it('clears moderators on unmount', () => {
    const clearModeratorSuggestions = jest.fn()
    const wrapper = shallow(<ModeratorsSettingsTab clearModeratorSuggestions={clearModeratorSuggestions} />)
    wrapper.unmount()
    expect(clearModeratorSuggestions).toHaveBeenCalled()
  })

  it('renders a list of RemovableListItems and AddModerator', () => {
    const moderators = [
      { id: 1 },
      { id: 2 },
      { id: 3 },
      { id: 4 }
    ]

    const wrapper = shallow(<ModeratorsSettingsTab moderators={moderators} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('calls removeModerator', () => {
    const props = {
      moderators: [
        { id: 1 },
        { id: 2 },
        { id: 3 },
        { id: 4 }
      ],
      slug: 'foogroup',
      removeModerator: jest.fn()
    }
    const wrapper = shallow(<ModeratorsSettingsTab {...props} />)

    wrapper.setState({ moderatorToRemove: 3, isRemoveFromGroup: false })
    wrapper.instance().submitRemoveModerator()
    expect(props.removeModerator).toHaveBeenCalledWith(3, false)
  })
})

describe('ModeratorsList', () => {
  it('renders correctly', () => {
    const props = {
      moderators: [
        { id: 1 },
        { id: 2 },
        { id: 3 },
        { id: 4 }
      ],
      slug: 'foogroup'
    }

    const wrapper = shallow(<ModeratorsList {...props} />)
    expect(wrapper).toMatchSnapshot()
  })
})

describe('AddModerator', () => {
  it('renders correctly, and transitions from not adding to adding', () => {
    const wrapper = shallow(<AddModerator />)
    expect(wrapper).toMatchSnapshot()
    wrapper.find('.add-new').simulate('click')
    expect(wrapper).toMatchSnapshot()
  })

  it('renders correctly when adding with suggestions', () => {
    const suggestions = [
      { id: 1, name: 'Demeter' },
      { id: 2, name: 'Ares' },
      { id: 1, name: 'Hermes' }
    ]
    const wrapper = shallow(<AddModerator moderatorSuggestions={suggestions} />)
    wrapper.find('.add-new').simulate('click')
    expect(wrapper).toMatchSnapshot()
  })

  it('handles interactions correctly', () => {
    const fetchModeratorSuggestions = jest.fn()
    const clearModeratorSuggestions = jest.fn()

    const wrapper = shallow(<AddModerator
      fetchModeratorSuggestions={fetchModeratorSuggestions}
      clearModeratorSuggestions={clearModeratorSuggestions} />)
    wrapper.find('.add-new').simulate('click')

    const input = wrapper.find('input')

    input.simulate('change', {
      target: { value: 'Artem' }
    })
    expect(fetchModeratorSuggestions).toHaveBeenCalledWith('Artem')
    expect(clearModeratorSuggestions).not.toHaveBeenCalled()

    fetchModeratorSuggestions.mockClear()
    clearModeratorSuggestions.mockClear()
    input.simulate('change', {
      target: { value: '' }
    })
    expect(clearModeratorSuggestions).toHaveBeenCalled()
    expect(fetchModeratorSuggestions).not.toHaveBeenCalled()

    fetchModeratorSuggestions.mockClear()
    clearModeratorSuggestions.mockClear()
    input.simulate('keyDown', {
      keyCode: keyMap.ENTER
    })

    clearModeratorSuggestions.mockClear()
    input.simulate('keyDown', {
      keyCode: keyMap.ENTER
    })
    expect(clearModeratorSuggestions).not.toHaveBeenCalled()

    clearModeratorSuggestions.mockClear()
    input.simulate('keyDown', {
      keyCode: keyMap.ESC
    })
    expect(clearModeratorSuggestions).toHaveBeenCalled()
  })
})
