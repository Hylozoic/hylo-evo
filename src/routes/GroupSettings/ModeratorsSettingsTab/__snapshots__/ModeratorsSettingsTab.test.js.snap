// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddModerator renders correctly when adding with suggestions 1`] = `
<div
  data-stylename="styles.adding"
>
  <div
    data-stylename="styles.help-text"
  >
    Search here for members to grant moderator powers
  </div>
  <div
    data-stylename="styles.input-row"
  >
    <input
      data-stylename="styles.input"
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="Type..."
      type="text"
    />
    <span
      data-stylename="styles.cancel-button"
      onClick={[Function]}
    >
      Cancel
    </span>
    <span
      data-stylename="styles.add-button"
      onClick={[Function]}
    >
      Add
    </span>
  </div>
  <div
    style={
      Object {
        "width": NaN,
      }
    }
  >
    <KeyControlledItemList
      items={
        Array [
          Object {
            "id": 1,
            "name": "Demeter",
          },
          Object {
            "id": 2,
            "name": "Ares",
          },
          Object {
            "id": 1,
            "name": "Hermes",
          },
        ]
      }
      onChange={[Function]}
      theme={Object {}}
    />
  </div>
</div>
`;

exports[`AddModerator renders correctly, and transitions from not adding to adding 1`] = `
<div
  data-stylename="styles.add-new"
  onClick={[Function]}
>
  + Add New
</div>
`;

exports[`AddModerator renders correctly, and transitions from not adding to adding 2`] = `
<div
  data-stylename="styles.adding"
>
  <div
    data-stylename="styles.help-text"
  >
    Search here for members to grant moderator powers
  </div>
  <div
    data-stylename="styles.input-row"
  >
    <input
      data-stylename="styles.input"
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="Type..."
      type="text"
    />
    <span
      data-stylename="styles.cancel-button"
      onClick={[Function]}
    >
      Cancel
    </span>
    <span
      data-stylename="styles.add-button"
      onClick={[Function]}
    >
      Add
    </span>
  </div>
</div>
`;

exports[`ModeratorsList renders correctly 1`] = `
<div>
  <div>
    <RemovableListItem
      item={
        Object {
          "id": 1,
        }
      }
      key="1"
      skipConfirm={true}
      url="/groups/foogroup/members/1"
    />
    <RemovableListItem
      item={
        Object {
          "id": 2,
        }
      }
      key="2"
      skipConfirm={true}
      url="/groups/foogroup/members/2"
    />
    <RemovableListItem
      item={
        Object {
          "id": 3,
        }
      }
      key="3"
      skipConfirm={true}
      url="/groups/foogroup/members/3"
    />
    <RemovableListItem
      item={
        Object {
          "id": 4,
        }
      }
      key="4"
      skipConfirm={true}
      url="/groups/foogroup/members/4"
    />
  </div>
  <AddModerator />
</div>
`;

exports[`ModeratorsSettingsTab renders a list of RemovableListItems and AddModerator 1`] = `
<Fragment>
  <SettingsSection>
    <h3>
      Moderators
      <div
        data-stylename="styles.help-text"
      >
        Who has access to group settings and moderation powers
      </div>
    </h3>
    <ModeratorsList
      key="mList"
      moderators={
        Array [
          Object {
            "id": 1,
          },
          Object {
            "id": 2,
          },
          Object {
            "id": 3,
          },
          Object {
            "id": 4,
          },
        ]
      }
      removeItem={[Function]}
    />
  </SettingsSection>
  <SettingsSection>
    <h3>
      Other Roles & Badges
    </h3>
    <div
      data-stylename="styles.help-text"
    >
      Create additional roles or badges for the group
    </div>
    <div
      data-stylename="styles.add-role"
      onClick={[Function]}
    >
      <h4>
        Create new role/badge
      </h4>
      <Icon
        data-stylename="styles.new-role"
        name="Circle-Plus"
      />
    </div>
  </SettingsSection>
  <br />
</Fragment>
`;

exports[`ModeratorsSettingsTab renders loading if no moderators 1`] = `<Loading />`;
